data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/buffer/KafkaBuffer.java

```diff
@@ -0,0 +1,91 @@
+package org.opensearch.dataprepper.plugins.kafka.buffer;
+
+import org.opensearch.dataprepper.metrics.PluginMetrics;
+import org.opensearch.dataprepper.model.CheckpointState;
+import org.opensearch.dataprepper.model.acknowledgements.AcknowledgementSetManager;
+import org.opensearch.dataprepper.model.annotations.DataPrepperPlugin;
+import org.opensearch.dataprepper.model.annotations.DataPrepperPluginConstructor;
+import org.opensearch.dataprepper.model.buffer.AbstractBuffer;
+import org.opensearch.dataprepper.model.buffer.Buffer;
+import org.opensearch.dataprepper.model.configuration.PluginSetting;
+import org.opensearch.dataprepper.model.plugin.PluginFactory;
+import org.opensearch.dataprepper.model.record.Record;
+import org.opensearch.dataprepper.plugins.buffer.blockingbuffer.BlockingBuffer;
+import org.opensearch.dataprepper.plugins.kafka.configuration.KafkaBufferConfig;
+import org.opensearch.dataprepper.plugins.kafka.consumer.KafkaCustomConsumer;
+import org.opensearch.dataprepper.plugins.kafka.consumer.KafkaCustomConsumerFactory;
+import org.opensearch.dataprepper.plugins.kafka.producer.KafkaCustomProducer;
+import org.opensearch.dataprepper.plugins.kafka.producer.KafkaCustomProducerFactory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+@DataPrepperPlugin(name = "kafka_buffer", pluginType = Buffer.class, pluginConfigurationType = KafkaBufferConfig.class)
+public class KafkaBuffer<T extends Record<?>> extends AbstractBuffer<T> {
+
+    private static final Logger LOG = LoggerFactory.getLogger(KafkaBuffer.class);
+    public static final int INNER_BUFFER_CAPACITY = 1000000;
+    public static final int INNER_BUFFER_BATCH_SIZE = 250000;
+    private final KafkaCustomProducer producer;
+    private final AbstractBuffer innerBuffer;
```

Comment:
There are a few things about the usage of `innerBuffer` that are very coupled with the implementation and could be prone to breaking as that class changes.

* Calling `doWrite` and `doRead`. This code should call the public methods to ensure that all important logic is used.
* Exposing `postProcess`. You have do do this because you are not calling `read` directly.

I think that you are trying to avoid creating duplicate metrics.

I think a better solution would be to use no-op metrics instead. Perhaps by passing in a `PluginMetrics` instance.

Reply:
I hope we are using minimal buffers and minimal copy in end to end packet path of this buffer plugin. Else it will bite us in OOM and performance.

Reply:
Is the innerbuffer same as existing memory buffer? writeAll from source would write to Kafka (using Kafka Producer API) and the consumer of Kafka would write to the memory buffer (as if the writeAll was writing directly to the memory buffer). The memory buffer should be same as the existing memory buffer and we should be able to configure it's size, just like the way we do it today